acceptancetests/deploy_stack.py

```diff
@@ -570,6 +571,40 @@ def tear_down(self, has_controller):
             self.tear_down_client.kill_controller(check=True)
 
 
+class ExistingController:
+    """A Controller strategy where the controller is already present.
+
+    Intended for use with BootstrapManager.
+    """
+
+    def __init__(self, client, tear_down_client, juju_data_dir, env):
+        self.client = client
+        self.client.env.environment = env
+        self.tear_down_client = tear_down_client
+
+    def create_initial_model(self, upload_tools):
+        """Create the initial model."""
+        self.client.add_model(self.client.env)
+        logging.info('Added model {} to existing controller'.format(
+            self.client.env.environment))
+
+    def prepare(self):
+        """Prepare client for use by killing the existing controller."""
+        self.tear_down_client.kill_controller()
+
+    def get_hosts(self):
+        """Provide the controller host."""
+        host = get_machine_dns_name(
+            self.client.get_controller_client(), '0')
+        if host is None:
+            raise ValueError('Could not get machine 0 host')
+        return {'0': host}
+
+    def tear_down(self, _):
+        """Model already destroyed, switch back to a known default"""
+        self.client.switch('controller')
```

Comment:
The switch is a bit of a quirk with the way client_for_existing() in version_client.py creates the initial client object. Without changing too much of the code the method needs an active model or it will barf. Since the default behavior for juju is to unset the current model after one is destroyed it won't work if we run any test in serial on an existing controller without this. I've spoken with @nskaggs about it a bit, he thinks it's an unneeded optimization. I'd like to hear your thoughts.

As for the cleanup, it's a context and logging thing. The model is destroyed when the context around it ends, which makes collecting logs from those machines difficult. Right now the code cleans up the model post-log collection as expected (such as on an exception).